backend/lib/src/services/auth/axum.rs

```diff
@@ -0,0 +1,189 @@
+    ///
+    /// The user is identified by the address used during the login flow
+    Authenticated { address: String },
+
+    /// Represents an unauthenticated user
+    ///
+    /// The user is identified by this ID, which holds no guarantees in terms of "stickiness" nor "uniqueness"
+    Unauthenticated { id: String },
+}
+
+enum AuthenticationResult {
+    Success(User),
+    NoJWT,
+    BadJWT(JwtClaims, Error),
+    Error(Error),
+}
+
+impl User {
+    /// Will return a string usable to identify the user for the session
+    ///
+    /// WARNING: Do not use for identify verification
+    pub fn id(&self) -> &String {
+        match self {
+            User::Authenticated { address } => &address,
+            User::Unauthenticated { id } => &id,
+        }
+    }
+
+    /// Will return the authenticated user address or error if the user is unauthenticated
+    pub fn address(&self) -> Result<&String, Error> {
+        match self {
+            Self::Authenticated { address } => Ok(&address),
+            _ => Err(Error::Unauthorized("User not authenticated".to_owned())),
+        }
+    }
+
+    async fn unauthenticated_from_request_parts<S>(
+        _parts: &mut Parts,
+        _state: &S,
+    ) -> Result<Self, Error>
+    where
+        S: Send + Sync,
+    {
+        // TODO: determine user ID from peer information in parts
+        Ok(Self::Unauthenticated {
+            id: "unauthenticated".to_string(),
+        })
+    }
+
+    /// Verifies the passed in `JwtClaims`
+    ///
+    /// Returns the user information if the claims are valid and not expired
+    // TODO: user logout verification
+    fn from_claims(claims: &JwtClaims) -> Result<Self, Error> {
+        let now = Utc::now();
+        let exp = DateTime::<Utc>::from_timestamp(claims.exp, 0)
+            .ok_or_else(|| Error::Unauthorized("Invalid JWT expiry".to_string()))?;
+        let iat = DateTime::<Utc>::from_timestamp(claims.iat, 0)
+            .ok_or_else(|| Error::Unauthorized("Invalid JWT issuance time".to_string()))?;
+
+        if now >= exp {
+            return Err(Error::Unauthorized("Expired JWT".to_string()));
+        }
+
+        if iat > now {
+            return Err(Error::Unauthorized("JWT issued in the future".to_string()));
+        }
+
+        Ok(Self::Authenticated {
+            address: claims.address.clone(),
+        })
+    }
+
+    async fn authenticated_from_request_parts<S>(
+        parts: &mut Parts,
+        state: &S,
+    ) -> AuthenticationResult
+    where
+        S: Send + Sync,
+        Decoder: FromRef<S>,
+    {
+        // We try to parse the JWT token, if not preset we proceed as Unauthenticated
+        let claims = Claims::<JwtClaims>::from_request_parts(parts, state).await;
+
+        match claims {
+            Ok(claims) => Self::from_claims(&claims.0)
+                .map(AuthenticationResult::Success)
+                .unwrap_or_else(|e| AuthenticationResult::BadJWT(claims.0, e)),
```

Comment:
Its confusing that an successful authentication can be `Sucess(User)`, that at the same time can be an `Unauthenticated`

Reply:
Because `User` has 2 variants... I guess we could rename this `AuthenticationResult` to `IdentificationResult` and change the semantics fo the operation...

In general, it will be up to the endpoint to decide which version they want for a given operation, if they are ok with just an unauthenticated user of if they want an actual address